shared/trieutil/sparse_merkle.go

```diff
@@ -10,64 +10,109 @@ import (
-	}
-	if index == len(m.originalItems) {
-		m.originalItems = append(m.originalItems, item)
-		return m.updateTrie()
-	}
-
-	m.originalItems[index] = item
-	return m.updateTrie()
-}
-
 // GenerateTrieFromItems constructs a Merkle trie from a sequence of byte slices.
-func GenerateTrieFromItems(items [][]byte, depth int) (*MerkleTrie, error) {
+func GenerateTrieFromItems(items [][]byte, depth int) (*SparseMerkleTrie, error) {
 	if len(items) == 0 {
 		return nil, errors.New("no items provided to generate Merkle trie")
 	}
-	layers := calcTreeFromLeaves(items, depth)
-	return &MerkleTrie{
+	leaves := items
+	layers := make([][][]byte, depth+1)
+	transformedLeaves := make([][]byte, len(leaves))
+	for i := range leaves {
+		arr := bytesutil.ToBytes32(leaves[i])
+		transformedLeaves[i] = arr[:]
+	}
+	layers[0] = transformedLeaves
+	for i := 0; i < depth; i++ {
+		if len(layers[i])%2 == 1 {
+			layers[i] = append(layers[i], zeroHashes[i])
+		}
+		updatedValues := make([][]byte, 0, 0)
+		for j := 0; j < len(layers[i]); j += 2 {
+			concat := hashutil.Hash(append(layers[i][j], layers[i][j+1]...))
+			updatedValues = append(updatedValues, concat[:])
+		}
+		layers[i+1] = updatedValues
+	}
+	return &SparseMerkleTrie{
 		branches:      layers,
 		originalItems: items,
 		depth:         uint(depth),
 	}, nil
 }
 
 // Items returns the original items passed in when creating the Merkle trie.
-func (m *MerkleTrie) Items() [][]byte {
+func (m *SparseMerkleTrie) Items() [][]byte {
 	return m.originalItems
 }
 
 // Root returns the top-most, Merkle root of the trie.
-func (m *MerkleTrie) Root() [32]byte {
+func (m *SparseMerkleTrie) Root() [32]byte {
 	enc := [32]byte{}
 	binary.LittleEndian.PutUint64(enc[:], uint64(len(m.originalItems)))
 	return hashutil.Hash(append(m.branches[len(m.branches)-1][0], enc[:]...))
 }
 
+// InsertIntoTrie inserts an item(deposit hash) into the trie.
+func (m *SparseMerkleTrie) InsertIntoTrie(item []byte, index int) {
+	for index >= len(m.branches[0]) {
+		m.branches[0] = append(m.branches[0], zeroHashes[0])
+	}
+	someItem := bytesutil.ToBytes32(item)
+	m.branches[0][index] = someItem[:]
+	if index >= len(m.originalItems) {
+		m.originalItems = append(m.originalItems, someItem[:])
+	} else {
+		m.originalItems[index] = someItem[:]
+	}
+	currentIndex := index
+	root := bytesutil.ToBytes32(item)
+	for i := 0; i < int(m.depth); i++ {
+		isLeft := currentIndex%2 == 0
+		neighborIdx := currentIndex ^ 1
+		neighbor := make([]byte, 32)
+		if neighborIdx >= len(m.branches[i]) {
+			neighbor = zeroHashes[i][:]
+		} else {
+			neighbor = m.branches[i][neighborIdx]
+		}
+		if isLeft {
+			parentHash := hashutil.Hash(append(root[:], neighbor...))
+			root = parentHash
+		} else {
+			parentHash := hashutil.Hash(append(neighbor, root[:]...))
+			root = parentHash
+		}
+		parentIdx := currentIndex / 2
+		// Update the cached layers at the parent index.
+		if len(m.branches[i+1]) == 0 || parentIdx >= len(m.branches[i+1]) {
```

Comment:
We need some special handling here since the trie is sparse and there may not be an actual object at the index we're trying to modify